server/lib/src/idm/application.rs

```diff
@@ -0,0 +1,741 @@
+use crate::prelude::*;
+use concread::cowcell::*;
+use hashbrown::HashMap;
+use kanidm_proto::internal::OperationError;
+use std::sync::Arc;
+use uuid::Uuid;
+
+#[derive(Clone)]
+pub(crate) struct Application {
+    pub uuid: Uuid,
+    pub name: String,
+    pub linked_group: Option<Uuid>,
+}
+
+impl Application {
+    #[cfg(test)]
+    pub(crate) fn try_from_entry_ro(
+        value: &Entry<EntrySealed, EntryCommitted>,
+        _qs: &mut QueryServerReadTransaction,
+    ) -> Result<Self, OperationError> {
+        if !value.attribute_equality(Attribute::Class, &EntryClass::Application.to_partialvalue()) {
+            return Err(OperationError::InvalidAccountState(
+                "Missing class: application".to_string(),
+            ));
+        }
+
+        let uuid = value.get_uuid();
+
+        let name = value
+            .get_ava_single_iname(Attribute::Name)
+            .map(|s| s.to_string())
+            .ok_or_else(|| {
+                OperationError::InvalidAccountState(format!(
+                    "Missing attribute: {}",
+                    Attribute::Name
+                ))
+            })?;
+
+        let linked_group: Option<Uuid> = value.get_ava_single_refer(Attribute::LinkedGroup);
+
+        Ok(Application {
+            name,
+            uuid,
+            linked_group,
+        })
+    }
+}
+
+#[derive(Clone)]
+struct LdapApplicationsInner {
+    set: HashMap<String, Application>,
+}
+
+pub struct LdapApplications {
+    inner: CowCell<LdapApplicationsInner>,
+}
+
+pub struct LdapApplicationsReadTransaction {
+    inner: CowCellReadTxn<LdapApplicationsInner>,
+}
+
+pub struct LdapApplicationsWriteTransaction<'a> {
+    inner: CowCellWriteTxn<'a, LdapApplicationsInner>,
+}
+
+impl<'a> LdapApplicationsWriteTransaction<'a> {
+    pub fn reload(&mut self, value: Vec<Arc<EntrySealedCommitted>>) -> Result<(), OperationError> {
+        let app_set: Result<HashMap<_, _>, _> = value
+            .into_iter()
+            .map(|ent| {
+                if !ent.attribute_equality(Attribute::Class, &EntryClass::Application.into()) {
+                    error!("Missing class application");
+                    return Err(OperationError::InvalidEntryState);
+                }
+
+                let uuid = ent.get_uuid();
+                let name = ent
+                    .get_ava_single_iname(Attribute::Name)
+                    .map(str::to_string)
+                    .ok_or(OperationError::InvalidValueState)?;
+                let linked_group = ent.get_ava_single_refer(Attribute::LinkedGroup);
+
+                let app = Application {
+                    uuid,
+                    name: name.clone(),
+                    linked_group,
+                };
+
+                Ok((name, app))
+            })
+            .collect();
+
+        app_set.map(|mut app_set| {
+            // Delay getting the inner mut (which may clone) until we know we are ok.
```

Comment:
Minor, but we might consider an addition to cowcell such that we can do "update_mut" so that we don't need to clone before we swap: https://github.com/kanidm/concread/blob/master/src/cowcell/mod.rs#L183

Reply:
https://github.com/kanidm/concread/pull/116/files

Added this for you, will update the concread dep in kanidm shortly.

Reply:
Great, rebased and changed to use .replace().